File: fixtures/scale/kotlin/medium/src/calc/parser/Parser.kt
package calc.parser

import calc.ast.BinaryExpr
import calc.ast.Expr
import calc.lexer.Token
import calc.lexer.TokenType

class Parser(tokens: List<Token>) {
    internal val cursor = TokenCursor(tokens)

    fun parseExpression(): Expr {
        val root = parseBinary(LOWEST)
        cursor.expect(TokenType.END)
        return root
    }

    internal fun parseBinary(minPrecedence: Int): Expr {
        var left = parsePrimary(this)
        while (precedenceOf(cursor.peek().type) > minPrecedence) {
            val op = cursor.advance()
            left = BinaryExpr(op.text[0], left, parseBinary(nextLevelFor(op.type)))
        }
        return left
    }
}

File: fixtures/scale/dart/medium/lib/parser/parser.dart
import '../ast/binary_expr.dart';
import '../ast/expr.dart';
import '../lexer/token.dart';
import '../lexer/token_type.dart';
import 'operands.dart';
import 'precedence.dart';
import 'token_cursor.dart';

class Parser {
  final TokenCursor cursor;

  Parser(List<Token> tokens) : cursor = TokenCursor(tokens);

  Expr parseExpression() {
    final root = parseBinary(lowest);
    cursor.expect(TokenType.end);
    return root;
  }

  Expr parseBinary(int minPrecedence) {
    var left = parsePrimary(this);
    while (precedenceOf(cursor.peek().type) > minPrecedence) {
      final op = cursor.advance();
      left = BinaryExpr(op.text, left, parseBinary(nextLevelFor(op.type)));
    }
    return left;
  }
}
